File: src/main/java/com/networkedassets/plugins/addpagetree/PageTreeService.java
package com.networkedassets.plugins.addpagetree;

import com.atlassian.confluence.pages.Page;
import com.atlassian.confluence.pages.PageManager;
import com.atlassian.confluence.security.Permission;
import com.atlassian.confluence.security.PermissionManager;
import com.atlassian.confluence.spaces.Space;
import com.atlassian.confluence.spaces.SpaceManager;
import com.atlassian.confluence.user.AuthenticatedUserThreadLocal;
import com.atlassian.sal.api.pluginsettings.PluginSettings;
import com.atlassian.sal.api.pluginsettings.PluginSettingsFactory;
import com.google.common.collect.Lists;
import com.networkedassets.plugins.addpagetree.managepages.Command;
import com.networkedassets.plugins.addpagetree.managepages.ExecutionContext;
import org.codehaus.jackson.map.ObjectMapper;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.ws.rs.*;
import javax.ws.rs.core.Response;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

@Path("/")
public class PageTreeService {
    private static final String LAST_COMMANDS = "com.networkedassets.plugins.add-pagetree.LAST_COMMANDS";
    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
    @SuppressWarnings("unused")
    private static Logger log = LoggerFactory.getLogger(PageTreeService.class);
    private PageManager pageManager;
    private PermissionManager permissionManager;
    private SpaceManager spaceManager;
    private PluginSettingsFactory pluginSettingsFactory;

    public PageTreeService(PageManager pageManager, PermissionManager permissionManager, SpaceManager spaceManager,
                           PluginSettingsFactory pluginSettingsFactory) {
        this.pageManager = pageManager;
        this.permissionManager = permissionManager;
        this.spaceManager = spaceManager;
        this.pluginSettingsFactory = pluginSettingsFactory;
    }

    private static Response error(String message) {
        return Response.status(500).entity(new JsonMessage(500, message)).build();
    }

    private static Response error(Exception exception) {
        return error(exception.getLocalizedMessage());
    }

    private static Response success(String message) {
        return Response.ok(new JsonMessage(200, message)).build();
    }

    @POST
    @Path("manage")
    @Produces({"application/json"})
    @Consumes({"application/json"})
    public Response managePages(@QueryParam("space") String space, List<Command> managePagesCommand) {
        Space s = spaceManager.getSpace(space);

        if (isUnauthorized(s)) return error("Unauthorized");
        if (managePagesCommand == null) return error("Did not get the modification commands");

        try {
            final Response error = executeCommands(s, managePagesCommand);
            if (error != null) return error;
        } catch (final IOException e) {
            return error(e);
        }

        return success("Success");
    }

    @Nullable
    private Response executeCommands(Space s, List<Command> commands) throws IOException {
        final ExecutionContext ec = new ExecutionContext(permissionManager, s);
        final List<Command> executedCommands = new ArrayList<>(commands.size());
        for (Command command : commands) {
            try {
                command.execute(pageManager, ec);
            } catch (final Exception e) {
                revertCommands(executedCommands, s);
                return error(e);
            }
            executedCommands.add(command);
        }
        persistLastChanges(executedCommands, ec);
        return null;
    }

    @POST
    @Path("revertLast")
    @Produces({"application/json"})
    @Consumes({"application/json"})
    public Response revertLastChange(@QueryParam("space") String spaceKey) {
        try {
            Space s = spaceManager.getSpace(spaceKey);
            LastChanges lastChanges = getLastChanges(s);
            if (lastChanges == null) return error("No last changes");
            if (!AuthenticatedUserThreadLocal.get().getKey().getStringValue().equals(lastChanges.userKey))
                return error("Unauthorized");
            List<Command> lastCommandList = lastChanges.executedCommands;
            revertCommands(lastCommandList, s);
        } catch (IOException e) {
            return error(e);
        }
        return success("Success");
    }

    private void persistLastChanges(List<Command> executedCommands, ExecutionContext ec) throws IOException {
        PluginSettings pluginSettings = pluginSettingsFactory.createSettingsForKey(ec.getSpace().getKey());
        LastChanges lc = new LastChanges();
        lc.executedCommands = executedCommands;
        lc.userKey = AuthenticatedUserThreadLocal.get().getKey().getStringValue();
        String serializedCommands = OBJECT_MAPPER.writeValueAsString(lc);
        pluginSettings.put(LAST_COMMANDS, serializedCommands);
    }

    @Nullable
    private LastChanges getLastChanges(Space space) throws IOException {
        PluginSettings pluginSettings = pluginSettingsFactory.createSettingsForKey(space.getKey());
        String serializedCommands = (String) pluginSettings.get(LAST_COMMANDS);
        try {
            return OBJECT_MAPPER.readValue(serializedCommands, LastChanges.class);
        } catch (IOException e) {
            return null;
        }
    }

    private void revertCommands(List<Command> executedCommands, Space s) throws IOException {
        for (Command executedCommand : Lists.reverse(executedCommands)) {
            executedCommand.revert(pageManager);
        }
        persistLastChanges(new ArrayList<>(), new ExecutionContext(permissionManager, s));
    }

    @GET
    @Path("pagetree")
    @Produces({"application/json"})
    public Response getPageTree(@QueryParam("space") String spaceKey, @QueryParam("rootPageId") Long rootId) {
        Space space = spaceManager.getSpace(spaceKey);
        if (isUnauthorized(space)) return error("Unauthorized");

        Page page = pageFromSpaceOrHomepage(rootId, space);

        final boolean canCreate = permissionManager.hasCreatePermission(AuthenticatedUserThreadLocal.get(), space, Page.class);
        PageTreeInfo pageTreeInfo = new PageTreeInfo(canCreate, JsonPage.from(page,
                AuthenticatedUserThreadLocal.get(),
                permissionManager));

        return Response.ok(pageTreeInfo).build();
    }

    private Page pageFromSpaceOrHomepage(Long pageId, Space space) {
        Page page;
        if (pageId == null || (page = pageManager.getPage(pageId)) == null || page.getSpace() != space)
            return space.getHomePage();
        else return page;
    }

    private boolean isUnauthorized(Space space) {
        return permissionManager == null || !permissionManager
                .hasPermission(AuthenticatedUserThreadLocal.get(), Permission.VIEW, space);
    }

    @SuppressWarnings("WeakerAccess")
    static class LastChanges {
        public List<Command> executedCommands;
        public String userKey;
    }

    @SuppressWarnings("WeakerAccess") // has to be public for jackson to be happy
    public static class PageTreeInfo {
        public boolean canCreate;
        public JsonPage pageTree;

        public PageTreeInfo(boolean canCreate, JsonPage pageTree) {
            this.canCreate = canCreate;
            this.pageTree = pageTree;
        }
    }

    @SuppressWarnings("WeakerAccess") // has to be public for jackson to be happy
    public static class JsonMessage {
        public int status;
        public String message;

        public JsonMessage(int status, String message) {
            this.status = status;
            this.message = message;
        }
    }
}

File: src/main/java/com/networkedassets/plugins/managemytree/PageTreeService.kt
package com.networkedassets.plugins.managemytree

import com.atlassian.confluence.pages.Page
import com.atlassian.confluence.pages.PageManager
import com.atlassian.confluence.security.Permission
import com.atlassian.confluence.security.PermissionManager
import com.atlassian.confluence.spaces.Space
import com.atlassian.confluence.spaces.SpaceManager
import com.atlassian.confluence.user.AuthenticatedUserThreadLocal
import com.atlassian.sal.api.pluginsettings.PluginSettingsFactory
import com.google.common.collect.Lists
import org.codehaus.jackson.annotate.JsonCreator
import org.codehaus.jackson.annotate.JsonProperty
import org.codehaus.jackson.map.ObjectMapper
import org.slf4j.LoggerFactory
import java.io.IOException
import java.util.*
import javax.ws.rs.*
import javax.ws.rs.core.Response

private val LAST_COMMANDS = "com.networkedassets.plugins.add-pagetree.LAST_COMMANDS"
private val OBJECT_MAPPER = ObjectMapper()
private val log = LoggerFactory.getLogger(PageTreeService::class.java)

@Path("/")
class PageTreeService(
        private val pageManager: PageManager,
        private val permissionManager: PermissionManager,
        private val spaceManager: SpaceManager,
        private val pluginSettingsFactory: PluginSettingsFactory
) {

    @POST
    @Path("manage")
    @Produces("application/json")
    @Consumes("application/json")
    fun managePages(@QueryParam("space") space: String, managePagesCommand: List<Command>?): Response {
        val s = spaceManager.getSpace(space)

        if (isUnauthorized(s)) return error("Unauthorized")
        if (managePagesCommand == null) return error("Did not get the modification commands")

        try {
            val error = executeCommands(s, managePagesCommand)
            if (error != null) return error
        } catch (e: IOException) {
            return error(e)
        }

        return success("Success")
    }

    private fun executeCommands(s: Space, commands: List<Command>): Response? {
        val ec = ExecutionContext(permissionManager, /*spaceBlueprintManager, blueprintContentGenerator,*/ s)
        val executedCommands = ArrayList<Command>(commands.size)
        for (command in commands) {
            try {
                command(pageManager, ec)
            } catch (e: Exception) {
                revertCommands(executedCommands, s)
                return error(e)
            }

            executedCommands.add(command)
        }
        persistLastChanges(executedCommands, ec)
        return null
    }

    @POST
    @Path("revertLast")
    @Produces("application/json")
    @Consumes("application/json")
    fun revertLastChange(@QueryParam("space") spaceKey: String): Response {
        try {
            val s = spaceManager.getSpace(spaceKey)
            val lastChanges = getLastChanges(s) ?: return error("No last changes")
            if (AuthenticatedUserThreadLocal.get().key.stringValue != lastChanges.userKey)
                return error("Unauthorized")
            val lastCommandList = lastChanges.executedCommands
            revertCommands(lastCommandList, s)

            return success("Success")
        } catch (e: IOException) {
            return error(e)
        }
    }

    private fun persistLastChanges(executedCommands: List<Command>, ec: ExecutionContext) {
        val pluginSettings = pluginSettingsFactory.createSettingsForKey(ec.space.key)
        val lc = LastChanges(executedCommands, AuthenticatedUserThreadLocal.get().key.stringValue)
        val serializedCommands = OBJECT_MAPPER.writeValueAsString(lc)
        pluginSettings.put(LAST_COMMANDS, serializedCommands)
    }

    private fun getLastChanges(space: Space): LastChanges? {
        val pluginSettings = pluginSettingsFactory.createSettingsForKey(space.key)
        val serializedCommands = pluginSettings.get(LAST_COMMANDS) as? String ?: return null
        return OBJECT_MAPPER.readValue(serializedCommands, LastChanges::class.java)
    }

    private fun revertCommands(executedCommands: List<Command>, s: Space) {
        for (executedCommand in Lists.reverse(executedCommands)) {
            executedCommand.revert(pageManager)
        }
        persistLastChanges(ArrayList<Command>(), ExecutionContext(permissionManager, /*spaceBlueprintManager, blueprintContentGenerator,*/ s))
    }

    @GET
    @Path("pagetree")
    @Produces("application/json")
    fun getPageTree(@QueryParam("space") spaceKey: String, @QueryParam("rootPageId") rootId: Long?): Response {
//        println("YO!!!!!")
        try {
            val space = spaceManager.getSpace(spaceKey)
            if (isUnauthorized(space)) return error("Unauthorized")

            val page = pageFromSpaceOrHomepage(rootId, space)
            val canCreate = permissionManager.hasCreatePermission(AuthenticatedUserThreadLocal.get(), space, Page::class.java)

            val lastChanges: LastChanges?
            var lastCommands: List<Command> = ArrayList()
            lastChanges = getLastChanges(space)
            if (lastChanges != null && AuthenticatedUserThreadLocal.get().key.stringValue == lastChanges.userKey)
                lastCommands = lastChanges.executedCommands

            val pageTreeInfo = PageTreeInfo(
                    canCreate,
                    JsonPage.from(
                            page,
                            AuthenticatedUserThreadLocal.get(),
                            permissionManager),
                    lastCommands)

            return Response.ok(OBJECT_MAPPER.writeValueAsString(pageTreeInfo)).build()
        } catch (e: Exception) {
            log.error("Exception: ", e)
            return error(e)
        }
    }

    private fun pageFromSpaceOrHomepage(pageId: Long?, space: Space): Page {
        val page = pageManager.getPage(pageId ?: return space.homePage)
        if (page == null || page.space != space) return space.homePage
        return page
    }

    private fun isUnauthorized(space: Space): Boolean {
        return !permissionManager.hasPermission(AuthenticatedUserThreadLocal.get(), Permission.VIEW, space)
    }
}

data class LastChanges
@JsonCreator constructor(
        @field:JsonProperty @param:JsonProperty("executedCommands") var executedCommands: List<Command>,
        @field:JsonProperty @param:JsonProperty("userKey") var userKey: String
)

data class PageTreeInfo
@JsonCreator constructor(
        @field:JsonProperty @param:JsonProperty("canCreate") var canCreate: Boolean,
        @field:JsonProperty @param:JsonProperty("pageTree") var pageTree: JsonPage,
        @field:JsonProperty @param:JsonProperty("lastChanges") var lastChanges: List<Command>
)

data class JsonMessage
@JsonCreator constructor(
        @field:JsonProperty @param:JsonProperty("status") var status: Int,
        @field:JsonProperty @param:JsonProperty("message") var message: String
)

fun error(message: String) = Response.status(500).entity(JsonMessage(500, message)).build()
fun error(exception: Exception) = error(exception.message ?: "")
fun success(message: String) = Response.ok(JsonMessage(200, message)).build()
